### ISSUE TITLE
Views/Working With Forms has a confusing (potentially wrong) statement about using for vs foreach

### ISSUE THREAD

**Issue by artiomchi** [aspnet/Docs#3863]
A statement on the [Working With Forms](https://docs.microsoft.com/en-us/aspnet/core/mvc/views/working-with-forms) has been brought up to me, and it got me a bit confused. 

It states (it's the only reference to `foreach` on the whole page):
> Always use `for` (and not `foreach`) to iterate over a list. Evaluating an indexer in a LINQ expression can be expensive and should be minimized.

Now this is interesting for two reasons.

First of all - wouldn't it be the other way around? If you want to iterate through the whole list, wouldn't foreach use an IEnumerator, which should be more efficient, when going through all the elements in a collection in order?

Secondly.. If evaluating indexers are expensive, shouldn't we **avoid** using a for loop?

It seems like this is either badly worded, and just adds confusion.. Or it's just plain wrong, in which case it needs amending :)

Thoughts?

**Comment by Rick-Anderson** [comment]
@NTaylorMullen please review

**Comment by NTaylorMullen** [comment]
So technically both arguments are correct, it matters where you utilize the indexed item. So, in general `for` is better than `foreach` (if the scenario allows it) because it doesn't need to allocate an enumerator. However, in the articles case it's bringing up how a foreach is better to use due to how its indexed item value is consumed in an `asp-for` scenario (it's simpler to evaluate).

tl;dr; The article could be fixed to say that foreach should be used if possible when the value is going to be used in an `asp-for` or `Html.DisplayFor` equivalent context.

**Comment by artiomchi** [comment]
@NTaylorMullen well, that's not right either, is it?

Consider the extract from there:
```cshtml
@for (int i = 0; i < Model.Count; i++)
{
    <tr>
        @Html.EditorFor(model => model[i])
    </tr>
}
```

You would have to use the for indexer in this case to make sure that each editor uses a different prefix. Otherwise all values would be posted under the same name.

One could argue that a foreach could be used for display fields, but we **have** have to use a for loop to render editors.

Even if the recommendation is to use the for loop for everything, the note in the docs could definitely be improved, since the current description is at best confusing or misleading, and at worst just seems incorrect.
